File: agent/component.py
import threading
import logging
from abc import ABC, abstractmethod
from typing import Optional

from event.event import Event
from event.event_type import EventType
from event.event_dispatcher import EventDispatcher
from event.event_queue import EventQueue
from event.event_consumer import EventConsumer

class Component(EventConsumer, ABC):
    def __init__(self, event_dispatcher: EventDispatcher, agent):
        self.thread: Optional[threading.Thread] = None
        self.event_dispatcher = event_dispatcher
        self.event_queue = EventQueue()
        self.agent = agent
        self.logger = logging.getLogger(self.__class__.__name__)
        self.subscribe_events()
    
    def start(self):
        self.thread = threading.Thread(target=self.event_handling_loop)
        self.thread.daemon = True
        self.thread.name = f"{self.__class__.__name__} Event Handler"
        self.thread.start()
    
    def terminate(self):
        if self.thread:
            self.thread.join(timeout=1.0)
    
    def set_up(self):
        """Override in subclasses to initialize the component"""
        pass
    
    @abstractmethod
    def subscribe_events(self):
        """Subscribe to events that this component wants to receive"""
        pass
    
    @abstractmethod
    def handle_event(self, event: Event) -> bool:
        """
        Process an event
        
        Returns:
            bool: True if the event was handled, False if it should be put back in the queue
        """
        pass
    
    def event_handling_loop(self):
        self.set_up()
        while True:
            try:
                event = self.event_queue.get()
                if not self.handle_event(event):
                    self.event_queue.put(event)
                self.event_queue.task_done()
            except Exception as e:
                self.logger.error(f"An error occurred while handling an event: {e}", exc_info=True)
    
    def receive_event(self, event: Event):
        self.event_queue.put(event)
    
    def dispatch_event(self, event_type: EventType, event_data=None):
        class SimpleEvent(Event):
            def get_type(self):
                return event_type
            
            def get_data(self):
                return event_data
        
        self.event_dispatcher.dispatch(SimpleEvent())
    
    def subscribe_event(self, event_type: EventType):
        self.event_dispatcher.add_subscription(event_type, self)

File: lapras-middleware/src/main/java/kr/ac/kaist/cdsn/lapras/Component.java
package kr.ac.kaist.cdsn.lapras;

import kr.ac.kaist.cdsn.lapras.event.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Created by Daekeun Lee on 2016-11-11.
 */
public abstract class Component implements EventConsumer {
    private static final Logger LOGGER = LoggerFactory.getLogger(Component.class);

    private Thread thread;
    private final EventDispatcher eventDispatcher;
    protected final EventQueue eventQueue;
    protected final Agent agent;

    public Component(EventDispatcher eventDispatcher, Agent agent) {
        this.eventDispatcher = eventDispatcher;
        this.agent = agent;
        this.eventQueue = new EventQueue();
        subscribeEvents();
    }

    public void start() {
        this.thread = new Thread(() -> eventHandlingLoop());
        this.thread.setName(String.format("%s Event Handler", this.getClass().getSimpleName()));
        this.thread.setDaemon(true);
        this.thread.start();
    }

    public void terminate() {
        this.thread.interrupt();
        try {
            this.thread.join();
        } catch (InterruptedException e) {
        }
    }

    public void setUp() {
    }

    protected abstract void subscribeEvents();

    protected abstract boolean handleEvent(Event event);

    private void eventHandlingLoop() {
        setUp();
        while(true) {
            try {
                Event event = this.eventQueue.take();
                if(!handleEvent(event)) {
                    this.eventQueue.put(event);
                }
            } catch (InterruptedException e) {
                break;
            } catch (Exception e) {
                LOGGER.error("An error occurred while handling an event", e);
            }
        }
    }

    @Override
    public void receiveEvent(Event e) {
        eventQueue.add(e);
    }

    protected void dispatchEvent(final EventType eventType, final Object eventData) {
        dispatchEvent(new Event() {
            @Override
            public EventType getType() {
                return eventType;
            }

            @Override
            public Object getData() {
                return eventData;
            }
        });
    }

    protected void dispatchEvent(Event event) {
        try {
            this.eventDispatcher.dispatch(event);
        } catch (InterruptedException e) {
        }
    }

    protected void subscribeEvent(EventType eventType) {
        eventDispatcher.addSubscription(eventType, this);
    }
}
